config/routes.rb

```diff
@@ -150,6 +150,15 @@
       patch :progress_for_clearance
       patch :flag_for_clearance
     end
+
+    member do
+      patch :mark_as_waiting_for_data, to: 'case_transitions#mark_as_waiting_for_data', as: 'mark_as_waiting_for_data'
+      patch :mark_as_ready_for_vetting, to: 'case_transitions#mark_as_ready_for_vetting', as: 'mark_as_ready_for_vetting'
+      patch :mark_as_vetting_in_progress, to: 'case_transitions#mark_as_vetting_in_progress', as: 'mark_as_vetting_in_progress'
+      patch :mark_as_ready_to_copy, to: 'case_transitions#mark_as_ready_to_copy', as: 'mark_as_ready_to_copy'
+      patch :mark_as_ready_to_dispatch, to: 'case_transitions#mark_as_ready_to_dispatch', as: 'mark_as_ready_to_dispatch'
+      patch :mark_as_closed, to: 'case_transitions#mark_as_closed', as: 'mark_as_closed'
+    end
```

Comment:
Ideally these routes are defined as part of `cases/offender_sar/mark_as_...`

Or if they must remain as a separate controller then there is no reason why the route cannot be something like:

```
/cases/:id/transitions/:transition_name 
```

So rather than defining each route, shis should allow the next route to be defined programatically as it is done. already by the state machine. Then in `cases_helper` rather than adding more `cases..when` can add a method like `CasesHelper#action_button_for_destroy_case` to generate this specific button. It should also mean the `#rubocop:disable Metrics/CyclomaticComplexity, Metrics/MethodLength` in `CasesHelper` can be removed (guess!).

Reply:
Maybe we could pair and you show me what you mean? I'm not quite clear from the above. I've moved the actions into offender_sar_controller. Presently the routes are 

```
                       mark_as_waiting_for_data_case PATCH  /cases/:id/mark_as_waiting_for_data(.:format)                           cases/offender_sar#mark_as_waiting_for_data
                      mark_as_ready_for_vetting_case PATCH  /cases/:id/mark_as_ready_for_vetting(.:format)                          cases/offender_sar#mark_as_ready_for_vetting
                    mark_as_vetting_in_progress_case PATCH  /cases/:id/mark_as_vetting_in_progress(.:format)                        cases/offender_sar#mark_as_vetting_in_progress
                          mark_as_ready_to_copy_case PATCH  /cases/:id/mark_as_ready_to_copy(.:format)                              cases/offender_sar#mark_as_ready_to_copy
                      mark_as_ready_to_dispatch_case PATCH  /cases/:id/mark_as_ready_to_dispatch(.:format)                          cases/offender_sar#mark_as_ready_to_dispatch
                                 mark_as_closed_case PATCH  /cases/:id/mark_as_closed(.:format)                                     cases/offender_sar#mark_as_closed
```

Using `resource :transition do` instead of `member do` they become
```
           mark_as_waiting_for_data_case_transitions PATCH  /cases/:case_id/transitions/mark_as_waiting_for_data(.:format)          cases/offender_sar#mark_as_waiting_for_data
          mark_as_ready_for_vetting_case_transitions PATCH  /cases/:case_id/transitions/mark_as_ready_for_vetting(.:format)         cases/offender_sar#mark_as_ready_for_vetting
        mark_as_vetting_in_progress_case_transitions PATCH  /cases/:case_id/transitions/mark_as_vetting_in_progress(.:format)       cases/offender_sar#mark_as_vetting_in_progress
              mark_as_ready_to_copy_case_transitions PATCH  /cases/:case_id/transitions/mark_as_ready_to_copy(.:format)             cases/offender_sar#mark_as_ready_to_copy
          mark_as_ready_to_dispatch_case_transitions PATCH  /cases/:case_id/transitions/mark_as_ready_to_dispatch(.:format)         cases/offender_sar#mark_as_ready_to_dispatch
                     mark_as_closed_case_transitions PATCH  /cases/:case_id/transitions/mark_as_closed(.:format)                    cases/offender_sar#mark_as_closed
```

When you suggest generating the routes programmatically, do you mean by making an array of the actions and looping over in the routes file, or something more cleverer?

Reply:
Amended using the approach you suggested, thanks!